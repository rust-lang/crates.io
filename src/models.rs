pub use self::action::{NewVersionOwnerAction, VersionAction, VersionOwnerAction};
pub use self::category::{Category, CrateCategory, NewCategory};
pub use self::crate_owner_invitation::{
    CrateOwnerInvitation, NewCrateOwnerInvitation, NewCrateOwnerInvitationOutcome,
};
pub use self::default_versions::{update_default_version, verify_default_version};
pub use self::deleted_crate::NewDeletedCrate;
pub use self::dependency::{Dependency, DependencyKind, ReverseDependency};
pub use self::download::VersionDownload;
pub use self::email::{Email, NewEmail};
pub use self::follow::Follow;
pub use self::keyword::{CrateKeyword, Keyword};
pub use self::krate::{Crate, CrateName, NewCrate, RecentCrateDownloads};
pub use self::owner::{CrateOwner, Owner, OwnerKind};
pub use self::rights::Rights;
pub use self::team::{NewTeam, Team};
pub use self::token::ApiToken;
pub use self::user::{NewUser, User};
pub use self::version::{NewVersion, TopVersions, Version};

pub mod helpers;

mod action;
pub mod category;
pub mod crate_owner_invitation;
pub mod default_versions;
mod deleted_crate;
pub mod dependency;
pub mod download;
mod email;
mod follow;
mod keyword;
pub mod krate;
mod owner;
mod rights;
mod team;
pub mod token;
pub mod user;
pub mod version;
