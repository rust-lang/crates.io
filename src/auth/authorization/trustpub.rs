use crate::auth::Permission;
use crate::util::errors::{BoxedAppError, forbidden};

pub struct AuthorizedTrustPub {
    crate_ids: Vec<i32>,
}

impl AuthorizedTrustPub {
    pub fn new(crate_ids: Vec<i32>) -> Self {
        Self { crate_ids }
    }

    pub(in crate::auth) async fn validate(
        self,
        permission: Permission<'_>,
    ) -> Result<Self, BoxedAppError> {
        let existing_crate = match permission {
            Permission::PublishUpdate { krate } => krate,
            Permission::PublishNew { .. } => {
                let message = "Trusted Publishing tokens do not support creating new crates";
                return Err(forbidden(message));
            }
            _ => {
                let message = "Trusted Publishing tokens can only be used for publishing crates";
                return Err(forbidden(message));
            }
        };

        if !self.crate_ids.contains(&existing_crate.id) {
            let name = &existing_crate.name;
            let error = format!("The provided access token is not valid for crate `{name}`");
            return Err(forbidden(error));
        }

        Ok(self)
    }
}
