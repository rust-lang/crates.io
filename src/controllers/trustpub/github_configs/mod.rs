pub mod create;
pub mod delete;
pub mod emails;
pub mod json;
