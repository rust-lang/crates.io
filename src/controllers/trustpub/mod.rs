pub mod github_configs;
pub mod tokens;
