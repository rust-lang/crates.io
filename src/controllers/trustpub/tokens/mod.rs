pub mod exchange;
pub mod json;
