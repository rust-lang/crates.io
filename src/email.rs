use crate::config;
use crate::Env;
use lettre::address::Envelope;
use lettre::message::header::ContentType;
use lettre::message::Mailbox;
use lettre::transport::file::AsyncFileTransport;
use lettre::transport::smtp::authentication::{Credentials, Mechanism};
use lettre::transport::smtp::AsyncSmtpTransport;
use lettre::transport::stub::AsyncStubTransport;
use lettre::{Address, AsyncTransport, Message, Tokio1Executor};
use rand::distributions::{Alphanumeric, DistString};
use std::sync::Arc;

pub trait Email {
    fn subject(&self) -> String;
    fn body(&self) -> String;
}

#[derive(Debug, Clone)]
pub struct Emails {
    backend: EmailBackend,
    pub domain: String,
    from: Address,
}

const DEFAULT_FROM: &str = "noreply@crates.io";

impl Emails {
    /// Create a new instance detecting the backend from the environment. This will either connect
    /// to a SMTP server or store the emails on the local filesystem.
    pub fn from_environment(config: &config::Server) -> Self {
        let login = dotenvy::var("MAILGUN_SMTP_LOGIN");
        let password = dotenvy::var("MAILGUN_SMTP_PASSWORD");
        let server = dotenvy::var("MAILGUN_SMTP_SERVER");

        let from = login.as_deref().unwrap_or(DEFAULT_FROM).parse().unwrap();

        let backend = match (login, password, server) {
            (Ok(login), Ok(password), Ok(server)) => {
                let transport = AsyncSmtpTransport::<Tokio1Executor>::relay(&server)
                    .unwrap()
                    .credentials(Credentials::new(login, password))
                    .authentication(vec![Mechanism::Plain])
                    .build();

                EmailBackend::Smtp(Box::new(transport))
            }
            _ => {
                let transport = AsyncFileTransport::new("/tmp");
                EmailBackend::FileSystem(Arc::new(transport))
            }
        };

        if config.base.env == Env::Production && !matches!(backend, EmailBackend::Smtp { .. }) {
            panic!("only the smtp backend is allowed in production");
        }

        let domain = config.domain_name.clone();

        Self {
            backend,
            domain,
            from,
        }
    }

    /// Create a new test backend that stores all the outgoing emails in memory, allowing for tests
    /// to later assert the mails were sent.
    pub fn new_in_memory() -> Self {
        Self {
            backend: EmailBackend::Memory(AsyncStubTransport::new_ok()),
            domain: "crates.io".into(),
            from: DEFAULT_FROM.parse().unwrap(),
        }
    }

    /// This is supposed to be used only during tests, to retrieve the messages stored in the
    /// "memory" backend. It's not cfg'd away because our integration tests need to access this.
    pub async fn mails_in_memory(&self) -> Option<Vec<(Envelope, String)>> {
        if let EmailBackend::Memory(transport) = &self.backend {
            Some(transport.messages().await)
        } else {
            None
        }
    }

    fn build_message(
        &self,
        recipient: &str,
        subject: String,
        body: String,
    ) -> Result<Message, EmailError> {
        // The message ID is normally generated by the SMTP server, but if we let it generate the
        // ID there will be no way for the crates.io application to know the ID of the message it
        // just sent, as it's not included in the SMTP response.
        //
        // Our support staff needs to know the message ID to be able to find misdelivered emails.
        // Because of that we're generating a random message ID, hoping the SMTP server doesn't
        // replace it when it relays the message.
        let message_id = format!(
            "<{}@{}>",
            Alphanumeric.sample_string(&mut rand::thread_rng(), 32),
            self.domain,
        );

        let from = Mailbox::new(Some(self.domain.clone()), self.from.clone());

        let message = Message::builder()
            .message_id(Some(message_id.clone()))
            .to(recipient.parse()?)
            .from(from)
            .subject(subject)
            .header(ContentType::TEXT_PLAIN)
            .body(body)?;

        Ok(message)
    }

    pub async fn send<E: Email>(&self, recipient: &str, email: E) -> Result<(), EmailError> {
        let email = self.build_message(recipient, email.subject(), email.body())?;

        self.backend
            .send(email)
            .await
            .map_err(EmailError::TransportError)
    }
}

#[derive(Debug, thiserror::Error)]
pub enum EmailError {
    #[error(transparent)]
    AddressError(#[from] lettre::address::AddressError),
    #[error(transparent)]
    MessageBuilderError(#[from] lettre::error::Error),
    #[error(transparent)]
    TransportError(anyhow::Error),
}

#[derive(Debug, Clone)]
enum EmailBackend {
    /// Backend used in production to send mails using SMTP.
    ///
    /// This is using `Box` to avoid a large size difference between variants.
    Smtp(Box<AsyncSmtpTransport<Tokio1Executor>>),
    /// Backend used locally during development, will store the emails in the provided directory.
    FileSystem(Arc<AsyncFileTransport<Tokio1Executor>>),
    /// Backend used during tests, will keep messages in memory to allow tests to retrieve them.
    Memory(AsyncStubTransport),
}

impl EmailBackend {
    async fn send(&self, message: Message) -> anyhow::Result<()> {
        match self {
            EmailBackend::Smtp(transport) => transport.send(message).await.map(|_| ())?,
            EmailBackend::FileSystem(transport) => transport.send(message).await.map(|_| ())?,
            EmailBackend::Memory(transport) => transport.send(message).await.map(|_| ())?,
        }

        Ok(())
    }
}

#[derive(Debug, Clone)]
pub struct StoredEmail {
    pub to: String,
    pub subject: String,
    pub body: String,
}

#[cfg(test)]
mod tests {
    use super::*;

    struct TestEmail;

    impl Email for TestEmail {
        fn subject(&self) -> String {
            "test".into()
        }

        fn body(&self) -> String {
            "test".into()
        }
    }

    #[tokio::test]
    async fn sending_to_invalid_email_fails() {
        let emails = Emails::new_in_memory();

        let address = "String.Format(\"{0}.{1}@live.com\", FirstName, LastName)";
        assert_err!(emails.send(address, TestEmail).await);
    }

    #[tokio::test]
    async fn sending_to_valid_email_succeeds() {
        let emails = Emails::new_in_memory();

        let address = "someone@example.com";
        assert_ok!(emails.send(address, TestEmail).await);
    }
}
