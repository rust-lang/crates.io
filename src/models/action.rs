use crate::models::{ApiToken, User, Version};
use crate::schema::*;
use bon::Builder;
use chrono::NaiveDateTime;
use crates_io_diesel_helpers::pg_enum;
use diesel::prelude::*;
use diesel_async::{AsyncPgConnection, RunQueryDsl};

pg_enum! {
    pub enum VersionAction {
        Publish = 0,
        Yank = 1,
        Unyank = 2,
    }
}

impl From<VersionAction> for &'static str {
    fn from(action: VersionAction) -> Self {
        match action {
            VersionAction::Publish => "publish",
            VersionAction::Yank => "yank",
            VersionAction::Unyank => "unyank",
        }
    }
}

impl From<VersionAction> for String {
    fn from(action: VersionAction) -> Self {
        let string: &'static str = action.into();

        string.into()
    }
}

#[derive(Debug, Clone, Copy, Queryable, Identifiable, Associations)]
#[diesel(
    table_name = version_owner_actions,
    check_for_backend(diesel::pg::Pg),
    belongs_to(Version),
    belongs_to(User, foreign_key = user_id),
    belongs_to(ApiToken, foreign_key = api_token_id),
)]
pub struct VersionOwnerAction {
    pub id: i32,
    pub version_id: i32,
    pub user_id: i32,
    pub api_token_id: Option<i32>,
    pub action: VersionAction,
    pub time: NaiveDateTime,
}

impl VersionOwnerAction {
    pub async fn all(conn: &mut AsyncPgConnection) -> QueryResult<Vec<Self>> {
        version_owner_actions::table.load(conn).await
    }

    pub async fn by_version(
        conn: &mut AsyncPgConnection,
        version: &Version,
    ) -> QueryResult<Vec<(Self, User)>> {
        use version_owner_actions::dsl::version_id;

        version_owner_actions::table
            .filter(version_id.eq(version.id))
            .inner_join(users::table)
            .order(version_owner_actions::dsl::id)
            .load(conn)
            .await
    }

    pub async fn for_versions(
        conn: &mut AsyncPgConnection,
        versions: &[&Version],
    ) -> QueryResult<Vec<Vec<(Self, User)>>> {
        Ok(Self::belonging_to(versions)
            .inner_join(users::table)
            .order(version_owner_actions::dsl::id)
            .load(conn)
            .await?
            .grouped_by(versions))
    }
}

#[derive(Insertable, Debug, Builder)]
#[diesel(table_name = version_owner_actions, check_for_backend(diesel::pg::Pg))]
pub struct NewVersionOwnerAction {
    version_id: i32,
    user_id: i32,
    api_token_id: Option<i32>,
    #[builder(into)]
    action: VersionAction,
}

impl NewVersionOwnerAction {
    pub async fn insert(&self, conn: &mut AsyncPgConnection) -> QueryResult<VersionOwnerAction> {
        diesel::insert_into(version_owner_actions::table)
            .values(self)
            .get_result(conn)
            .await
    }
}
