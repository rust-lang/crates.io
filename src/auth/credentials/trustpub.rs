use crate::auth::{AuthorizedTrustPub, Permission};
use crate::controllers;
use crate::util::errors::{BoxedAppError, custom, forbidden};
use axum::extract::FromRequestParts;
use crates_io_database::schema::trustpub_tokens;
use crates_io_trustpub::access_token::{AccessToken, AccessTokenError};
use diesel::dsl::now;
use diesel::{ExpressionMethods, OptionalExtension, QueryDsl};
use diesel_async::{AsyncPgConnection, RunQueryDsl};
use http::StatusCode;
use http::header::ToStrError;
use http::request::Parts;

#[derive(Debug)]
pub struct TrustPubCredentials {
    token: AccessToken,
}

impl TrustPubCredentials {
    pub fn from_request_parts(parts: &Parts) -> Result<Self, TrustPubCredentialsError> {
        use TrustPubCredentialsError::*;
        use http::header;

        let header = parts.headers.get(header::AUTHORIZATION);
        let header = header.ok_or(MissingAuthorizationHeader)?;
        let header = header.to_str()?;

        let (scheme, token) = header.split_once(' ').unwrap_or(("", header));
        if !(scheme.is_empty() || scheme.eq_ignore_ascii_case("Bearer")) {
            return Err(InvalidAuthScheme);
        }

        Self::from_raw_token(token.trim_ascii())
    }

    pub fn from_raw_token(token: &str) -> Result<Self, TrustPubCredentialsError> {
        let token = token.parse::<AccessToken>()?;
        Ok(Self { token })
    }

    pub async fn validate(
        &self,
        conn: &mut AsyncPgConnection,
        _parts: &Parts,
        permission: Permission<'_>,
    ) -> Result<AuthorizedTrustPub, BoxedAppError> {
        let hashed_token = self.token.sha256();

        let crate_ids = trustpub_tokens::table
            .filter(trustpub_tokens::hashed_token.eq(hashed_token.as_slice()))
            .filter(trustpub_tokens::expires_at.gt(now))
            .select(trustpub_tokens::crate_ids)
            .get_result::<Vec<Option<i32>>>(conn)
            .await
            .optional()?
            .ok_or_else(|| forbidden("Invalid authentication token"))?;

        let crate_ids = crate_ids.into_iter().flatten().collect();

        AuthorizedTrustPub::new(crate_ids)
            .validate(permission)
            .await
    }

    pub fn unvalidated_token(&self) -> &AccessToken {
        &self.token
    }
}

impl<S: Send + Sync> FromRequestParts<S> for TrustPubCredentials {
    type Rejection = BoxedAppError;

    async fn from_request_parts(parts: &mut Parts, _: &S) -> Result<Self, Self::Rejection> {
        controllers::util::verify_origin(parts)?;
        Ok(Self::from_request_parts(parts)?)
    }
}

#[derive(Debug, thiserror::Error)]
pub enum TrustPubCredentialsError {
    #[error("Missing `Authorization` header")]
    MissingAuthorizationHeader,
    #[error("Unexpected non-ASCII characters in `Authorization` header: {0}")]
    InvalidCharacters(#[from] ToStrError),
    #[error("Unexpected `Authorization` header scheme")]
    InvalidAuthScheme,
    #[error("Invalid access token: {0}")]
    InvalidAccessToken(#[from] AccessTokenError),
}

impl From<TrustPubCredentialsError> for BoxedAppError {
    fn from(err: TrustPubCredentialsError) -> Self {
        if matches!(err, TrustPubCredentialsError::InvalidAccessToken(_)) {
            let message = "Invalid `Authorization` header: Failed to parse token";
            custom(StatusCode::UNAUTHORIZED, message)
        } else if matches!(err, TrustPubCredentialsError::MissingAuthorizationHeader) {
            let message = "Missing `Authorization` header";
            custom(StatusCode::UNAUTHORIZED, message)
        } else {
            let message = format!("Authentication failed: {err}");
            custom(StatusCode::UNAUTHORIZED, message)
        }
    }
}
