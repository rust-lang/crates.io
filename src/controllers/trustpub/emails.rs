use crate::email::EmailMessage;
use crates_io_database::models::trustpub::{GitHubConfig, GitLabConfig};
use crates_io_database::models::{Crate, User};

#[derive(Debug, Clone, Copy, serde::Serialize)]
#[serde(tag = "type")]
pub enum ConfigType<'a> {
    GitHub(&'a GitHubConfig),
    GitLab(&'a GitLabConfig),
}

#[derive(serde::Serialize)]
pub struct ConfigCreatedEmail<'a> {
    /// The GitHub login of the email recipient.
    pub recipient: &'a str,
    /// The user who created the trusted publishing configuration.
    pub auth_user: &'a User,
    /// The crate for which the trusted publishing configuration was created.
    pub krate: &'a Crate,
    /// The trusted publishing configuration that was created.
    pub saved_config: ConfigType<'a>,
}

impl ConfigCreatedEmail<'_> {
    pub fn render(&self) -> Result<EmailMessage, minijinja::Error> {
        EmailMessage::from_template("trustpub_config_created", self)
    }
}

#[derive(serde::Serialize)]
pub struct ConfigDeletedEmail<'a> {
    /// The GitHub login of the email recipient.
    pub recipient: &'a str,
    /// The user who deleted the trusted publishing configuration.
    pub auth_user: &'a User,
    /// The crate for which the trusted publishing configuration was deleted.
    pub krate: &'a Crate,
    /// The trusted publishing configuration that was deleted.
    pub config: ConfigType<'a>,
}

impl ConfigDeletedEmail<'_> {
    pub fn render(&self) -> Result<EmailMessage, minijinja::Error> {
        EmailMessage::from_template("trustpub_config_deleted", self)
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use chrono::Utc;
    use claims::assert_ok;
    use insta::assert_snapshot;

    fn test_user() -> User {
        User {
            id: 1,
            gh_login: "octocat".into(),
            name: Some("The Octocat".into()),
            gh_id: 123,
            gh_avatar: None,
            gh_encrypted_token: vec![],
            account_lock_reason: None,
            account_lock_until: None,
            is_admin: false,
            publish_notifications: true,
        }
    }

    fn test_crate() -> Crate {
        Crate {
            id: 1,
            name: "my-crate".into(),
            updated_at: Utc::now(),
            created_at: Utc::now(),
            description: None,
            homepage: None,
            documentation: None,
            repository: None,
            max_upload_size: None,
            max_features: None,
        }
    }

    fn test_github_config(environment: Option<&str>) -> GitHubConfig {
        GitHubConfig {
            id: 1,
            created_at: Utc::now(),
            crate_id: 1,
            repository_owner_id: 42,
            repository_owner: "rust-lang".into(),
            repository_name: "rust".into(),
            workflow_filename: "publish.yml".into(),
            environment: environment.map(String::from),
        }
    }

    fn test_gitlab_config(environment: Option<&str>) -> GitLabConfig {
        GitLabConfig {
            id: 1,
            created_at: Utc::now(),
            crate_id: 1,
            namespace_id: None,
            namespace: "rust-lang".into(),
            project: "my-crate".into(),
            workflow_filepath: ".gitlab-ci.yml".into(),
            environment: environment.map(String::from),
        }
    }

    #[test]
    fn test_config_created_email() {
        let email = ConfigCreatedEmail {
            recipient: "octocat",
            auth_user: &test_user(),
            krate: &test_crate(),
            saved_config: ConfigType::GitHub(&test_github_config(None)),
        };

        let rendered = assert_ok!(email.render());
        assert_snapshot!(rendered.subject, @"crates.io: Trusted Publishing configuration added to my-crate");
        assert_snapshot!(rendered.body_text);
    }

    #[test]
    fn test_config_created_email_with_environment() {
        let email = ConfigCreatedEmail {
            recipient: "octocat",
            auth_user: &test_user(),
            krate: &test_crate(),
            saved_config: ConfigType::GitHub(&test_github_config(Some("production"))),
        };

        let rendered = assert_ok!(email.render());
        assert_snapshot!(rendered.subject, @"crates.io: Trusted Publishing configuration added to my-crate");
        assert_snapshot!(rendered.body_text);
    }

    #[test]
    fn test_config_created_email_different_recipient() {
        let email = ConfigCreatedEmail {
            recipient: "team-member",
            auth_user: &test_user(),
            krate: &test_crate(),
            saved_config: ConfigType::GitHub(&test_github_config(None)),
        };

        let rendered = assert_ok!(email.render());
        assert_snapshot!(rendered.subject, @"crates.io: Trusted Publishing configuration added to my-crate");
        assert_snapshot!(rendered.body_text);
    }

    #[test]
    fn test_config_created_email_gitlab() {
        let email = ConfigCreatedEmail {
            recipient: "octocat",
            auth_user: &test_user(),
            krate: &test_crate(),
            saved_config: ConfigType::GitLab(&test_gitlab_config(None)),
        };

        let rendered = assert_ok!(email.render());
        assert_snapshot!(rendered.subject, @"crates.io: Trusted Publishing configuration added to my-crate");
        assert_snapshot!(rendered.body_text);
    }

    #[test]
    fn test_config_created_email_gitlab_with_environment() {
        let email = ConfigCreatedEmail {
            recipient: "octocat",
            auth_user: &test_user(),
            krate: &test_crate(),
            saved_config: ConfigType::GitLab(&test_gitlab_config(Some("production"))),
        };

        let rendered = assert_ok!(email.render());
        assert_snapshot!(rendered.subject, @"crates.io: Trusted Publishing configuration added to my-crate");
        assert_snapshot!(rendered.body_text);
    }

    #[test]
    fn test_config_deleted_email() {
        let email = ConfigDeletedEmail {
            recipient: "octocat",
            auth_user: &test_user(),
            krate: &test_crate(),
            config: ConfigType::GitHub(&test_github_config(None)),
        };

        let rendered = assert_ok!(email.render());
        assert_snapshot!(rendered.subject, @"crates.io: Trusted Publishing configuration removed from my-crate");
        assert_snapshot!(rendered.body_text);
    }

    #[test]
    fn test_config_deleted_email_with_environment() {
        let email = ConfigDeletedEmail {
            recipient: "octocat",
            auth_user: &test_user(),
            krate: &test_crate(),
            config: ConfigType::GitHub(&test_github_config(Some("production"))),
        };

        let rendered = assert_ok!(email.render());
        assert_snapshot!(rendered.subject, @"crates.io: Trusted Publishing configuration removed from my-crate");
        assert_snapshot!(rendered.body_text);
    }

    #[test]
    fn test_config_deleted_email_different_recipient() {
        let email = ConfigDeletedEmail {
            recipient: "team-member",
            auth_user: &test_user(),
            krate: &test_crate(),
            config: ConfigType::GitHub(&test_github_config(None)),
        };

        let rendered = assert_ok!(email.render());
        assert_snapshot!(rendered.subject, @"crates.io: Trusted Publishing configuration removed from my-crate");
        assert_snapshot!(rendered.body_text);
    }

    #[test]
    fn test_config_deleted_email_gitlab() {
        let email = ConfigDeletedEmail {
            recipient: "octocat",
            auth_user: &test_user(),
            krate: &test_crate(),
            config: ConfigType::GitLab(&test_gitlab_config(None)),
        };

        let rendered = assert_ok!(email.render());
        assert_snapshot!(rendered.subject, @"crates.io: Trusted Publishing configuration removed from my-crate");
        assert_snapshot!(rendered.body_text);
    }

    #[test]
    fn test_config_deleted_email_gitlab_with_environment() {
        let email = ConfigDeletedEmail {
            recipient: "octocat",
            auth_user: &test_user(),
            krate: &test_crate(),
            config: ConfigType::GitLab(&test_gitlab_config(Some("production"))),
        };

        let rendered = assert_ok!(email.render());
        assert_snapshot!(rendered.subject, @"crates.io: Trusted Publishing configuration removed from my-crate");
        assert_snapshot!(rendered.body_text);
    }
}
