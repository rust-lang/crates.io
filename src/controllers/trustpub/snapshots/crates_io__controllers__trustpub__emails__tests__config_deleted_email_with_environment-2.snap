---
source: src/controllers/trustpub/emails.rs
expression: rendered.body_text
---
Hello octocat!

You removed a "Trusted Publishing" configuration for GitHub Actions from your crate "my-crate".

The removed configuration was for the workflow file at https://github.com/rust-lang/rust/blob/HEAD/.github/workflows/publish.yml using the `production` environment.

If you did not make this change and you think it was made maliciously, you can email help@crates.io for assistance.

--
The crates.io Team
